File: linkedlist/Q138.java
import java.util.HashMap;
import java.util.Map;

public class Q138 {
    public Node copyRandomList1(Node head) {
        if (head == null) return null;

        Node ptr = head, next;

        while (ptr != null) {
            next = ptr.next;
            ptr.next = new Node(ptr.val);
            ptr.next.next = next;
            ptr = next;
        }

        ptr = head;
        while (ptr != null) {
            next = ptr.next.next;
            if (ptr.random != null) {
                ptr.next.random = ptr.random.next;
            }
            ptr = next;
        }

        ptr = head;
        Node dummy = new Node(0), cur = dummy;
        while (ptr != null) {
            cur.next = ptr.next;
            cur = cur.next;
            ptr.next = ptr.next.next;
            ptr = ptr.next;
        }
        return dummy.next;
    }

    public Node copyRandomList2(Node head) {
        if (head == null) return null;

        Map<Node, Node> oldToNew = new HashMap<>();

        for (Node ptr = head; ptr != null; ptr = ptr.next) {
            oldToNew.put(ptr, new Node(ptr.val));
        }

        for (Node ptr = head; ptr != null; ptr = ptr.next) {
            oldToNew.get(ptr).next = oldToNew.get(ptr.next);
            oldToNew.get(ptr).random = oldToNew.get(ptr.random);
        }
        return oldToNew.get(head);
    }
}


class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

File: linkedlist/Q138.py
# https://leetcode.com/problems/copy-list-with-random-pointer/

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        # Use hash table, two pass

        if not head:
            return None

        oldToNew = dict()

        # 1. shadow copy
        ptr = head
        while ptr:
            oldToNew[ptr] = Node(ptr.val, ptr.next, ptr.random)
            ptr = ptr.next

        # 2. repace next and random
        ptr = head
        while ptr:
            if ptr.next:
                oldToNew[ptr].next = oldToNew[oldToNew[ptr].next]
            if ptr.random:
                oldToNew[ptr].random = oldToNew[oldToNew[ptr].random]
            ptr = ptr.next

        return oldToNew[head]

    def copyRandomList(self, head: 'Node') -> 'Node':
        # No extra space, no hash table

        if not head:
            return None

        # 1. shadow copy
        ptr = head
        while ptr:
            oldNext = ptr.next
            copy = Node(ptr.val, oldNext)
            ptr.next = copy
            ptr = oldNext

        # 2. copy actual random pointer
        ptr = head
        while ptr:
            if ptr.random:
                ptr.next.random = ptr.random.next
            ptr = ptr.next.next

        # 3. restore original list and construct new list
        newHead = head.next
        ptr = head
        while ptr:
            oldNext = ptr.next.next
            newNext = ptr.next
            newNext.next = oldNext.next if oldNext else None
            ptr.next = oldNext
            ptr = oldNext
        return newHead
